src/content/blog/2025-05-19-state-of-open-source-ai-powered-test-automation.md

```diff
@@ -0,0 +1,362 @@
+        EC.element_to_be_clickable((By.XPATH, '/html/body/div/div/div[2]/button[14]'))
+    )
+    browser.execute_script('arguments[0].click();', element)
+
+    element = WebDriverWait(browser, 10).until(
+        EC.element_to_be_clickable((By.XPATH, '/html/body/div/div/div[2]/button[16]'))
+    )
+    browser.execute_script('arguments[0].click();', element)
+
+    element = WebDriverWait(browser, 10).until(
+        EC.element_to_be_clickable((By.XPATH, '/html/body/div/div/div[2]/button[14]'))
+    )
+    browser.execute_script('arguments[0].click();', element)
+
+    element = WebDriverWait(browser, 10).until(
+        EC.element_to_be_clickable((By.XPATH, '/html/body/div/div/div[2]/button[20]'))
+    )
+    browser.execute_script('arguments[0].click();', element)
+
+@then('the calculator value should be 4')
+def the_calculator_value_should_be_4(browser: webdriver.Chrome):
+
+    from selenium.webdriver.common.by import By
+    from selenium.webdriver.support.ui import WebDriverWait
+    from selenium.webdriver.support import expected_conditions as EC
+
+    # Fixture for expected calculator value
+    expected_calculator_value = "4"
+
+    # Locate the calculator output element
+    calculator_output_element = WebDriverWait(browser, 10).until(
+        EC.presence_of_element_located((By.XPATH, "//div[@class='output']/input[@id='output']"))
+    )
+
+    # Get the actual value from the calculator output
+    actual_calculator_value = calculator_output_element.get_attribute("value").strip()
+
+    # Assert that the actual calculator value matches the expected value
+    assert actual_calculator_value == expected_calculator_value
+```
+
+</details>
+
+That’s a lot of code for such a simple test! Nevertheless, we can clearly see following parts:
+
+- browser is started (lines 15-20);
+- web application is opened (lines 23-25);
+- "2+2=" is performed (lines 26-46);
+- check that calculator result is 4 (lines 48-67).
+
+Of course, the code is not perfect and unlikely to be written by a seasoned test engineer. It relies on brittle hierarchical XPath, mixes implicit and explicit waits, has expected conditions that are not necessary, and plenty of duplication. Still, the code is working and can be refactored (by LLM too). Let’s run it to make sure it’s fine:
+
+```bash
+❱ pytest -q ./generated_tests/calculator.py
+.                                                                        [100%]
+1 passed in 2.14s
+```
+
+Not too bad: it’s a reliable and fast test (2 seconds)!
+
+In a nutshell, test generators work great for new projects without any test automation, but once the project matures and I want to re-use the code between tests and scale them, the re-generation and further reworking of the tests would take more and more time. Plus, there is no automated re-generator process for UI updates. With the current state of AI-powered agentic coding tools, the same can be achieved by crafting a proper prompt in Cursor/Claude Code/etc.
+
+**Pros**
```

Comment:
```suggestion
**Pros:**
```

Reply:
wdyt of adding an h3 header for the "pros and cons" part so they'll pop up in the Table of Contents?